server/src/test/kotlin/fi/oph/ludos/certificate/CertificateControllerTest.kt

```diff
@@ -0,0 +1,214 @@
+        assertEquals("Suko Certificate content Fi updated", updatedCertificate.description)
+        assertEquals(PublishState.PUBLISHED, updatedCertificate.publishState)
+        assertEquals("updated_certificate.pdf", updatedCertificate.fileName)
+        assertEquals("https://amazon_url.com/updated_certificate.pdf", updatedCertificate.fileKey)
+        assertNotNull(updatedCertificate.fileUploadDate)
+        assertEquals(certificateIn.id, updatedCertificate.id)
+    }
+
+    @Test
+    @WithYllapitajaRole
+    fun failCertificateUpdate() {
+        val nonExistentId = -1
+
+
+        val editedCertificate = """
+            {
+                "id": "$nonExistentId",
+                "exam": "SUKO",
+                "name": "New test name",
+                "description": "content",
+                "publishState": "PUBLISHED",
+                "fileName": "updated_certificate.pdf",
+                "fileKey": "https://amazon_url.com/updated_certificate.pdf",
+                "fileUploadDate": "${ZonedDateTime.now(ZoneOffset.UTC)}"
+            }
+        """.trimMargin()
+
+        val failUpdate =
+            mockMvc.perform(updateCertificate(nonExistentId, editedCertificate)).andReturn().response.contentAsString
+
+        assertEquals("Certificate not found $nonExistentId", failUpdate)
+    }
+
+    @Test
+    @WithYllapitajaRole
+    fun invalidExam() {
+        val body = """
+            {
+                "name": "Test Certificate",
+                "description": "Certificate content",
+                "publishState": "PUBLISHED",
+                "exam": "WRONG",
+                "fileName": "test_certificate.pdf",
+                "fileKey": "https://amazon_url.com/test_certificate.pdf",
+                "fileUploadDate": "2023-06-13"
+            }
+        """
+
+        mockMvc.perform(postCertificate(body)).andExpect(status().isBadRequest()).andReturn()
+    }
+
+    @Test
+    @WithYllapitajaRole
+    fun invalidState() {
+        val body = """
+        {
+            "name": "Test Certificate",
+            "description": "Certificate content",
+            "publishState": "",
+            "exam": "SUKO",
+            "fileName": "test_certificate.pdf",
+            "fileKey": "https://amazon_url.com/test_certificate.pdf",
+            "fileUploadDate": "2023-06-13"
+        }
+        """
+
+        val postResult = mockMvc.perform(postCertificate(body)).andExpect(status().isBadRequest()).andReturn()
+        val responseContent = postResult.response.contentAsString
+
+        assertThat(responseContent).contains("Invalid JSON payload: JSON parse error")
+    }
+
+    @Test
+    @WithYllapitajaRole
+    fun certificateNotFound() {
+        val getResult = mockMvc.perform(getCertificateById(Exam.SUKO, 999)).andExpect(status().isNotFound()).andReturn()
+        val responseContent = getResult.response.contentAsString
+
+        assertEquals(responseContent, "Certificate not found 999")
+    }
```

Comment:
Nyt ei oo erillistä getCertificateById-testiä, jossa haku onnistuu. Mut ehkä se riittää, et se testataan muissa testeissä.

Lisäisin käyttöoikeustestit jokaiseen endpointiin, esim. et ope ei voi luoda todistusta. Näis tietoturvajutuis on hyvä olla vyö ja henkselit, jos vaik vahingos unohtaa tai poistaa rooli-annotaation endpointista.

Reply:
Ja lisää testejä:
- uploadille ja previewille vielä kattavat testit mukaanlukien käyttöoikeuscheckit
- create ja update, jos fileKey ei oo olemassa kannassa. Näistä pitäis varmaan tulla 400 sit.